src/SandboxingLevels.php

```diff
@@ -0,0 +1,180 @@
+ */
+
+namespace AmpProject\AmpWP;
+
+use AMP_Comments_Sanitizer;
+use AMP_Form_Sanitizer;
+use AMP_Options_Manager;
+use AMP_Script_Sanitizer;
+use AmpProject\AmpWP\Infrastructure\Conditional;
+use AmpProject\AmpWP\Infrastructure\Registerable;
+use AmpProject\AmpWP\Infrastructure\Service;
+
+/**
+ * Experimental service to facilitate flexible AMP.
+ *
+ * @package AmpProject\AmpWP
+ * @since 2.2
+ * @internal
+ */
+final class SandboxingLevels implements Service, Registerable, Conditional {
+
+	/**
+	 * Option key for sandboxing level.
+	 *
+	 * @todo Move this to the Options interface once no longer experimental.
+	 * @var string
+	 */
+	const OPTION_SANDBOXING_LEVEL = 'sandboxing_level';
+
+	/**
+	 * Sandboxing levels.
+	 *
+	 * @var int[]
+	 */
+	const SANDBOXING_LEVELS = [ 1, 2, 3 ];
+
+	/**
+	 * Default sandboxing level.
+	 *
+	 * @var int
+	 */
+	const DEFAULT_SANDBOXING_LEVEL = 3;
+
+	/**
+	 * Whether service is needed.
+	 *
+	 * @return bool
+	 */
+	public static function is_needed() {
+		/**
+		 * Filters whether experimental sandboxing is enabled.
+		 *
+		 * @internal
+		 * @since 2.2
+		 *
+		 * @param bool $enabled Sandboxing enabled.
+		 */
+		return (bool) apply_filters( 'amp_experimental_sandboxing_enabled', false );
+	}
+
+	/**
+	 * Register.
+	 */
+	public function register() {
+		add_filter( 'amp_rest_options_schema', [ $this, 'filter_rest_options_schema' ] );
+		add_filter( 'amp_default_options', [ $this, 'filter_default_options' ], 10, 2 );
+		add_filter( 'amp_options_updating', [ $this, 'sanitize_options' ], 10, 2 );
+
+		add_action( 'init', [ $this, 'add_hooks' ] );
+	}
+
+	/**
+	 * Filter the REST options schema to add items.
+	 *
+	 * @param array $schema Schema.
+	 * @return array Schema.
+	 */
+	public function filter_rest_options_schema( $schema ) {
+		return array_merge(
+			$schema,
+			[
+				self::OPTION_SANDBOXING_LEVEL => [
+					'type'    => 'int',
+					'enum'    => self::SANDBOXING_LEVELS,
+					'default' => self::DEFAULT_SANDBOXING_LEVEL,
+				],
+			]
+		);
+	}
+
+	/**
+	 * Add default options.
+	 *
+	 * @param array $defaults Default options.
+	 * @return array Defaults.
+	 */
+	public function filter_default_options( $defaults ) {
+		$defaults[ self::OPTION_SANDBOXING_LEVEL ] = self::DEFAULT_SANDBOXING_LEVEL;
+		return $defaults;
+	}
+
+	/**
+	 * Sanitize options.
+	 *
+	 * @param array $options     Existing options with already-sanitized values for updating.
+	 * @param array $new_options Unsanitized options being submitted for updating.
+	 * @return array Sanitized options.
+	 */
+	public function sanitize_options( $options, $new_options ) {
+		if (
+			isset( $new_options[ self::OPTION_SANDBOXING_LEVEL ] )
+			&&
+			in_array( $new_options[ self::OPTION_SANDBOXING_LEVEL ], self::SANDBOXING_LEVELS, true )
+		) {
+			$options[ self::OPTION_SANDBOXING_LEVEL ] = $new_options[ self::OPTION_SANDBOXING_LEVEL ];
+		}
+		return $options;
+	}
+
+	/**
+	 * Add hooks.
+	 */
+	public function add_hooks() {
```

Comment:
I'd probably turn this into a big switch statement based on level, to make it clearer and easier to extend over time.
This could then make use of intentional fall through as well where needed.